### ISSUE TITLE
M:N zone aware routing

### ISSUE THREAD

**Issue by derekargueta** [envoyproxy/envoy#6333]
*Title*: M:N zone aware routing

*Description*:
Currently zone aware routing is only supported if there are equal numbers of zones for both the client and server per the documentation. It'd improve flexibility if zone aware routing can occur between M client zones and N server zones, such as if a particular service fails in one zone, the client could still try to optimize routing locally.

This seems orthogonal to https://github.com/envoyproxy/envoy/issues/1796.

**Comment by mattklein123** [comment]
@derekargueta agreed it would be nice to have this work if someone wants to do the extra algorithm work.

**Comment by derekargueta** [comment]
@mattklein123 we have an internal implementation in Finagle that is quite similar to Envoy's, with additional handling for the cases where the number of zones mismatch.

@fishcakez @jparise thoughts on creating a publicly available version of the design and providing an implementation to Envoy?

cc @rgs1

**Comment by mattklein123** [comment]
@derekargueta FWIW I doubt this would be difficult to implement in Envoy, even without any example code open sourced. I think the math is pretty straightforward. Someone just needs to work on it. (Working code would be nice also though!)

**Comment by derekargueta** [comment]
It's not yet clear if we can provide code. Agreed it's fairly straightforward, but since I'm well-versed in our internal implementation I think I'll pass on pursuing this as a precaution, so for now consider this an open invitation for a volunteer 🙂

We _are_ looking at helping out with https://github.com/envoyproxy/envoy/issues/5598

**Comment by sayap** [comment]
Just notice that this has been fixed by https://github.com/envoyproxy/envoy/pull/28970
